client/state/signup/progress/reducer.js

```diff
@@ -3,132 +3,120 @@
+const overwriteSteps = ( state, { steps = {} } ) => keyBy( steps, 'stepName' );
 
-	return state.filter( step => flowSteps.includes( step.stepName ) );
-}
+const completeStep = ( state, { step } ) => updateStep( state, { ...step, status: 'completed' } );
 
-function saveStep( state, { step } ) {
-	if ( find( state, { stepName: step.stepName } ) ) {
-		return updateStep( state, step );
-	}
+const invalidateStep = ( state, { step, errors } ) => {
+	const newStepState = { ...step, errors, status: 'invalid' };
 
-	return addStep( state, { ...step, status: 'in-progress' } );
-}
+	return has( state, step.stepName )
+		? updateStep( state, newStepState )
+		: addStep( state, newStepState );
+};
+
+const processStep = ( state, { step } ) => updateStep( state, { ...step, status: 'processing' } );
+
+const saveStep = ( state, { step } ) =>
+	has( state, step.stepName )
+		? updateStep( state, step )
+		: addStep( state, { ...step, status: 'in-progress' } );
 
 function setResumeAfterLogin( state, { resumeStep } ) {
 	debug( `Setting resume after login for step ${ resumeStep.stepName }` );
 	return updateStep( state, { ...resumeStep, status: 'in-progress' } );
 }
 
-function submitStep( state, { step } ) {
+const submitStep = ( state, { step } ) => {
 	const stepHasApiRequestFunction = get( stepsConfig, [ step.stepName, 'apiRequestFunction' ] );
 	const status = stepHasApiRequestFunction ? 'pending' : 'completed';
 
-	if ( find( state, { stepName: step.stepName } ) ) {
-		return updateStep( state, { ...step, status } );
-	}
-
-	return addStep( state, { ...step, status } );
-}
+	return has( state, step.stepName )
+		? updateStep( state, { ...step, status } )
+		: addStep( state, { ...step, status } );
+};
 
-//
-// Helper Functions
-//
-function addStep( state, step ) {
-	debug( `Adding step ${ step.stepName }` );
-	return [ ...state, step ];
-}
-
-function updateStep( state, newStep ) {
-	debug( `Updating step ${ newStep.stepName }` );
-	return map( state, function( step ) {
-		if ( step.stepName === newStep.stepName ) {
-			const { status } = newStep;
-			if ( status === 'pending' || status === 'completed' ) {
-				// This can only happen when submitting a step
-				//
-				// Steps that are resubmitted may decide to omit the `wasSkipped` status of a step if e.g.
-				// the user goes back and chooses to not skip a step. If a step is submitted without it,
-				// we explicitly remove it from the step data.
-				const { wasSkipped, ...commonStepArgs } = step;
-				return { ...commonStepArgs, ...newStep };
-			}
-
-			return { ...step, ...newStep };
-		}
-
-		return step;
-	} );
-}
-
-//
-// Reducer export
-//
 export default createReducer(
-	[],
+	{},
 	{
 		[ SIGNUP_COMPLETE_RESET ]: overwriteSteps,
 		[ SIGNUP_PROGRESS_COMPLETE_STEP ]: completeStep,
 		[ SIGNUP_PROGRESS_INVALIDATE_STEP ]: invalidateStep,
 		[ SIGNUP_PROGRESS_PROCESS_STEP ]: processStep,
-		[ SIGNUP_PROGRESS_REMOVE_UNNEEDED_STEPS ]: removeUnneededSteps,
```

Comment:
This PR removes the reducer, but the constant and the action itself are still used in a few places. There are no other reducers handling this so it should be safe to remove any other references as well.

Reply:
By the way it's not a blocker for merging - it can totally be adressed in a subsequent PR

Reply:
Thanks for spotting this Adam! I've opened a follow up PR here: #35461
Once I finish up a test plan and once this is merged I'll publish as a proper PR :)